### ISSUE TITLE
Virtual Scrolling with Remote Data in Reactive Grid

### ISSUE THREAD

**Issue by SergeyAlexeev** [DevExpress/devextreme-reactive#1912]
# The Problem

The current Virtual Scrolling implementation requires that all data rows are provided simultaneously. It slows down the Grid at startup and generates an excessive amount of server traffic.

# The Proposed Solution

We are going to introduce a new 'VirtualTableState' plugin that manages grid rows.

The `getRows` property contains a function that uses the `skip` and `take` arguments to load new rows. It is called when a user scrolls the Grid to the bottom.

The `start` property allows you to scroll the Grid to a specific row. This row is rendered at the top of the table. Set this property to the index of the row in the entire dataset.

The `rowCount` property sets the total row count. The greater the `rowCount`, the smaller the vertical scrollbar. If the `rowCount` is unspecified, pages are loaded in sequence, and the scrollbar's size depends on the number of rendered rows.

To reduce the number of requests to the server, the Grid will cache a portion of loaded rows. We will also provide utilities that help you configure row caching.

In addition, the Grid makes scrolling smoother by rendering skeleton cells for rows that should be loaded.

![grid-vs](https://user-images.githubusercontent.com/1897913/54419478-99679380-4718-11e9-814a-86295d34211e.gif)

```jsx
class MyApp extends React.Component {
  constructor() {
    /* ... */
    this.state = {
      rows: [],
      totalCount: 0,
      start: 0
    };
  }
  getRemoteRows(skip, take) {
    const query = `https://YOUR_API/?requireTotalCount=true&skip=${skip}&take=${take}`;
    fetch(query)
      .then(response => response.json())
      .then((data) => {
          this.setState({
            rows: data.data,
            totalCount: data.totalCount,
            start: skip,
          });
      });
  }
  render() {
    const { rows, start, totalCount } = this.state;
    return (
      <Grid rows={rows}>
        <VirtualTableState
          start={start}
          rowCount={totalCount}
          getRows={this.getRemoteRows}
        />
        <VirtualTable />
      </Grid>
    );
  }
}
```

# We Need Your Feedback

## Take a Quick Poll

[Do you find lazy loading in Reactive Grid useful?](https://docs.google.com/forms/d/e/1FAIpQLSevUBDDSJoPuvDA9kaUkP3ko-8_mDyuRD22sGz_0ae5QW41Kw/viewform?usp=sf_link)

## Live Sandbox

_Coming soon._

# Get Notified of Updates

Subscribe to this thread or to our [Facebook](https://www.facebook.com/DevExpress.DevExtreme/) and [Twitter](https://twitter.com/devextreme) accounts for updates on this topic.

**Comment by SergeyAlexeev** [comment]
Thank you for pointing out to that. We'll take your scenario into account.

**Comment by MaximKudriavtsev** [comment]
We have been released the `@next` version of [DevExtreme Reactive Grid](https://devexpress.github.io/devextreme-reactive/@next/react/grid/docs/guides/getting-started/) with this feature. Feel free to provide your feedback about its functionality and [documentation](https://devexpress.github.io/devextreme-reactive/@next/react/grid/docs/guides/lazy-loading/).

A package that contains the described functionality has the `next` tag:

```
npm i --save @devexpress/dx-react-core@next @devexpress/dx-react-grid@next
npm i --save @devexpress/dx-react-grid-material-ui@next
OR
npm i --save @devexpress/dx-react-grid-bootstrap3@next
OR
npm i --save @devexpress/dx-react-grid-bootstrap4@next
```

**Comment by MatthieuCoelho** [comment]
Nice feature thanks.
I tried to use it, homewer I wonder how to refresh the grid, I tried to do this  
```
this.cache.invalidate();
this.setState({rows: [], skip: 0, loading: true, totalCount: 5000});
```
but then getRows isn't called by the grid, why ?

**Comment by ushkal** [comment]
Hi @MatthieuCoelho,
 
Thank you for pointing out the issue. We'll fix it soon.

**Comment by ushkal** [comment]
@MatthieuCoelho, we researched your scenario and concluded that this behavior is expected: the grid requests rows only on user interactions such as scrolling, changing sorting, or filtering. In your case, I suggest you manually fetch rows on refresh as follows:

```
this.cache.invalidate();
this.getRemoteRows(this.state.skip, VIRTUAL_PAGE_SIZE * 2);
```

**Comment by MaximKudriavtsev** [comment]
We have released the `1.11.0` version of [DevExtreme Reactive Grid](https://devexpress.github.io/devextreme-reactive/react/grid/docs/guides/lazy-loading/) with this feature.

A package that contains the described functionality has the `latest` tag:

```
npm i --save @devexpress/dx-react-core @devexpress/dx-react-grid
npm i --save @devexpress/dx-react-grid-material-ui
OR
npm i --save @devexpress/dx-react-grid-bootstrap3
OR
npm i --save @devexpress/dx-react-grid-bootstrap4
```
